java/arcs/core/util/StateMachine.kt

```diff
@@ -0,0 +1,209 @@
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.util
+
+typealias EntryHandler<MACHINE> = suspend (MACHINE).() -> Unit
+typealias ErrorHandler<MACHINE> = suspend (MACHINE).(Exception) -> Unit
+
+/**
+ * [StateMachine] is an immutable directed graph of vertices representing states, and edges
+ * representing transitions. Each [STATE] can have an associated [onEntry], [onExit], [onError]
+ * handler and one or more [on] event transition handlers.
+ *
+ * [onEntry] handlers are invoked each time a [STATE] is entered, and [onExit] each time a state
+ * is left. [onError] is invoked if an exception is thrown by a handler.
+ *
+ * [on] event transition handlers specify which [EVENT] they trigger on, and the new [STATE] to
+ * transition to, effectively inserting an edge between the current state, and the new state
+ * into the graph.
+ *
+ * To trigger a transition, you [trigger] the [EVENT] associated with that edge. Events are queued,
+ * and then handlers are run in the following order:
+ * 1) [onExit] is invoked for the current state
+ * 2) If an exception occurs, [onError] is invoked and current and pending transitions are aborted
+ * 3) [on] event transition handler is invoked, if an exception occurs, [onError] is invoked
+ * and the transition and pending transitions are aborted.
+ * 4) The current state is set to the transition's new state.
+ * 5) The new state's [onEntry] handler is invoked.
+ * 6) If an exception occurs, [onError] is invoked.
+ * 7) Any [trigger] events that happen during these sequence are queued and processed as the last
+ * step. [onError] flushes pending events.
+ *
+ * [StateMachine] represents the only legal transitions possible. Triggering events
+ * which are not relevant for the current state results in an exception. Invalid state transition
+ * attempts are not silently ignored.
+ *
+ * Loosely inspired by those used in gaming or protocol frameworks, like Fettle
+ * [http://thehiflyer.github.io/Fettle/] but simplified for Arcs needs, and constructed via
+ * idiomatic Kotlin type-safe builders.
+ *
+ * @param STATE usually an enum or sealed class representing possible states
+ * @param EVENT usually an enum or sealed class representing possible event triggered transitions
+ * @param MACHINE the subclass type that extends [StateMachine]
+ * @property state the current/starting state of the machine.
+ */
+abstract class StateMachine<STATE, EVENT, MACHINE : StateMachine<STATE, EVENT, MACHINE>>(
+    var state: STATE
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
```

Comment:
Are we worried about thread-safety for this?

Might be worth considering Flow/Channel, as opposed to manually handling the queue.